Cargo cross compilation support

bonzini opened the pull request:
A fourth part of my Cargo interpreter fixes, adding cross compilation support for procedural macro crates and everything that they depend on.

~~A very early draft, I'll make sure it works on a few more projects before making it non-draft.~~

~~Note that machine selection has the same issue as feature selection: if a subproject has a dependency on crate A for the host and another subproject has a dependency on crate B for the build machine, picking them one after another with `subproject` will miss the `override_dependency` for the build machine and cross compilation will remain broken.~~

~~Fixing this requires either a two-phase approach where all subprojects are added before feature resolution (and `meson.build` generation) proceeds; or triggering subproject compilation from a `Cargo.toml` in the toplevel projects instead of having Cargo interpretation just for subprojects (#14639).~~

Fixes: #11744

bonzini committed:
cargo: vary crate names per machine

When a crate name is compiled for both the build and the host machine,
give it two different names to avoid conflicts.  This is not needed
for subprojects, but it is still necessary for packages that are part
of the toplevel workspace.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1946,6 +1946,8 @@ def _get_rust_crate_name(target_name: str) -> str:
         # Also +SUFFIX is dropped, which can be used to distinguish host from build crates
         crate_name = target_name.replace('-', '_').replace(' ', '_').replace('.', '_')
+        if '+' in crate_name:
+            crate_name = re.sub(r'\+.*$', '', crate_name)
```
Doh, the split one should have been obvious. Thanks!

bonzini committed:
cargo: implement machine-specific dependency tracking

Change the dependency resolution code to take a MachineChoice as an
argument.  Create build- and/or host-side configurations
through the machines_from() method in Manifest.

sp1ritCS commented:
Currently trying to cross-compile loupe and noticed some things:

- Dealing with git packages appears to be rather problematic right now. I don't know if this is due to the git package having its own meson.build file (which right now still just does custom target cargo wrapper); nor am I seemingly able to have a patch to it picked up packagefiles.

- I'm not sure why, but the zbus re-export of zvariant somehow requires, that dependents that depend on zbus not just need `--extern zbus=...` but also `--extern zvariant=...`, as otherwise you'll get

```
error[E0433]: failed to resolve: could not find `zvariant` in the list of imported crates
  --> ../subprojects/glycin/glycin-utils/src/api/loader.rs:37:73
   |
37 | #[cfg_attr(feature = "external", derive(DeserializeDict, SerializeDict, Type))]
   |                                                                         ^^^^ could not find `zvariant` in the list of imported crates
   |
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider importing one of these crates
   |
1  + use crate::zvariant;
   |
1  + use zbus::zvariant;
```

even though `loader.rs` contains `use zbus::zvariant::{self, DeserializeDict, Optional, SerializeDict, Type};`.

- I've encountered the following failure, somewhat related to the previous proc-macro issues:

```
[...]
zvariant_utils-3-rs| Rust linker for the build machine:  ld.bfd 2.45.0.20251103-2
zvariant_utils-3-rs| Message: Enabled features: ['default', 'gvariant']

Executing subproject ashpd-0.13-rs:zvariant-5-rs:zvariant_derive-5-rs:zvariant_utils-3-rs (build):serde-1-rs method cargo for machine: build

serde-1-rs| Generated Meson AST: /home/florian/workspace/Android/loupe/_cross/subprojects/serde-1.0.228/meson.build
serde-1-rs| Project name: serde-1.0.228 (for build machine)
serde-1-rs| Project version: 1.0.228
serde-1-rs| Rust compiler for the host machine: /home/florian/workspace/Android/AndroidRust2/linux-x86/1.88.0/bin/rustc --target aarch64-linux-android -C panic=abort -C linker=/home/florian/Android/Sdk/ndk/29.0.14033849/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android31-clang (rustc 1.88.0 "1.88.0-dev")
serde-1-rs| Rust linker for the host machine:  ld.lld 21.0.0
serde-1-rs| Rust compiler for the build machine: /home/florian/workspace/Android/AndroidRust2/linux-x86/1.88.0/bin/rustc -C panic=abort -C linker=cc (rustc 1.88.0 "1.88.0-dev")
serde-1-rs| Rust linker for the build machine:  ld.bfd 2.45.0.20251103-2
serde-1-rs| Message: Enabled features: ['default', 'derive', 'serde_derive', 'std']
serde-1-rs| Configuring private.rs using configuration

Executing subproject ashpd-0.13-rs:zvariant-5-rs:zvariant_derive-5-rs:zvariant_utils-3-rs (build):serde-1-rs:serde_core-1-rs method cargo for machine: build

serde_core-1-rs| Generated Meson AST: /home/florian/workspace/Android/loupe/_cross/subprojects/serde_core-1.0.228/meson.build
serde_core-1-rs| Project name: serde_core-1.0.228 (for build machine)
serde_core-1-rs| Project version: 1.0.228
serde_core-1-rs| Rust compiler for the host machine: /home/florian/workspace/Android/AndroidRust2/linux-x86/1.88.0/bin/rustc --target aarch64-linux-android -C panic=abort -C linker=/home/florian/Android/Sdk/ndk/29.0.14033849/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android31-clang (rustc 1.88.0 "1.88.0-dev")
serde_core-1-rs| Rust linker for the host machine:  ld.lld 21.0.0
serde_core-1-rs| Rust compiler for the build machine: /home/florian/workspace/Android/AndroidRust2/linux-x86/1.88.0/bin/rustc -C panic=abort -C linker=cc (rustc 1.88.0 "1.88.0-dev")
serde_core-1-rs| Rust linker for the build machine:  ld.bfd 2.45.0.20251103-2
serde_core-1-rs| Message: Enabled features: ['result', 'std']
serde_core-1-rs| Configuring private.rs using configuration
serde_core-1-rs| Build targets in project: 1143
serde_core-1-rs| Subproject serde_core-1-rs finished.


Executing subproject ashpd-0.13-rs:zvariant-5-rs:zvariant_derive-5-rs:zvariant_utils-3-rs (build):serde-1-rs:serde_derive-1-rs method cargo for machine: build

serde_derive-1-rs| Generated Meson AST: /home/florian/workspace/Android/loupe/_cross/subprojects/serde_derive-1.0.228/meson.build
serde_derive-1-rs| Project name: serde_derive-1.0.228 (for build machine)
serde_derive-1-rs| Project version: 1.0.228
serde_derive-1-rs| Rust compiler for the host machine: /home/florian/workspace/Android/AndroidRust2/linux-x86/1.88.0/bin/rustc --target aarch64-linux-android -C panic=abort -C linker=/home/florian/Android/Sdk/ndk/29.0.14033849/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android31-clang (rustc 1.88.0 "1.88.0-dev")
serde_derive-1-rs| Rust linker for the host machine:  ld.lld 21.0.0
serde_derive-1-rs| Rust compiler for the build machine: /home/florian/workspace/Android/AndroidRust2/linux-x86/1.88.0/bin/rustc -C panic=abort -C linker=cc (rustc 1.88.0 "1.88.0-dev")
serde_derive-1-rs| Rust linker for the build machine:  ld.bfd 2.45.0.20251103-2
serde_derive-1-rs| Message: Enabled features: ['default']

subprojects/serde_derive-1.0.228/meson.build:28:-1: ERROR: Tried to override dependency 'serde_derive-1-rs' which has already been resolved or overridden at /home/florian/workspace/Android/loupe/subprojects/serde_derive-1.0.228/Cargo.toml:28:
```

where the proc-macro override logic fails.

bonzini commented:
Thanks, I will take a look at the proc-macro issue. The git issue seems independent of cross compilation.

sp1ritCS commented:
If I understand it right, the proc-macro issue effectively occurs as the proc-macro subproject gets built both as a host and build subproject (from #15404) but both times actually building for the build machine.

Under the assumption, that a package providing a proc-macro will only provide a proc-macro, the following code works and lets me compile Loupe:

```diff
diff --git a/mesonbuild/modules/rust.py b/mesonbuild/modules/rust.py
index 52279723a..8e164425f 100644
--- a/mesonbuild/modules/rust.py
+++ b/mesonbuild/modules/rust.py
@@ -172,7 +172,7 @@ class RustWorkspace(ModuleObject):
             'options': None,
             'cmake_options': [],
             'default_options': {},
-            'for_machine': for_machine,
+            'for_machine': MachineChoice.BUILD if 'proc-macro' in pkg.supported_abis() else for_machine,
         }
         subp_name = pkg.get_subproject_name()
         self.interpreter.do_subproject(subp_name, kw, force_method='cargo')
```

bonzini commented:
Yes, that's the issue. Crates with a weird lib type can be handled too but it's great that such a small change is enough for the common case. I must be doing at least something right! :)

EDIT: added the fix to the top commit of #15404

bonzini committed:
cargo: assert that proc-macro crates are built once

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>

bonzini committed:
cargo: store machine-specific configuration

Make the PackageConfiguration a PerMachine object.  This makes it
possible to do dependency resolution separately for build- and/or host-side
libraries.

For now, all users only fill in or consume the host side, but that will change.
